### PR TITLE
Adapts to USRP X410

### PR DESCRIPTION
**Pull request by LawsonGu**
- **Summary**: 1. Make this API adapt to USRP X410 usage cases. 2. bug fix

- **Detailed Description**:  

  1. One difference between USRP X310 and X410 is Tx channel number from 2 to 4: to select Tx channels needs to add tx_duc_id, thus in Tx only script or main script it read this parameter. We need to add this parameter to all config file template. 

  2.  Decriptive content(like comment) change to let users know that we also support USRP X410
  
  3.   Fixed a bug which leads to users could not apply there RF parameteres(rate,gain, bandwidth,frequence) to the receive channel.
  4.  If target location of receiving not found, create a folder automatically.
  5. Add an animation when transmitting so that not seems task dead.

 

- **Affected Components**: Main scripts, Tx only, Rx only and all config templates are affected. 

- **Additional Testing/Validation**:  

       Using two USRP X410(1TX1RX) test system: 

image 

  

       On these usage cases and scope 

image 

  

- **Checklist**:  

  - [ √ ] Revert Scripts without functional changes. 

  - [ √ ] (Python) Reformat your code using Black. 

  - [ √ ] Document your code. 

  - [ √ ] Remove all debug code. 

  - [ √ ] Functional changes are tested.

### CODE REVIEW

**Review by mauro-belgiovine — APPROVED** [review]

**Comment by Abdo-Gaber** [comment]
The review is on going

File: README.md [hunk 1]
```diff
@@ -79,18 +79,19 @@ Despite being in the early phase of research for 6G, AI & ML appears to be becom
...
 ### Hardware
-To use the NI RF Data Recording API, you need at least one NI RF USRP device (for Tx or Rx only operation mode). The API has been tested only on [NI USRP X310](https://www.ettus.com/all-products/x310-kit/) ([X300/X310 Getting Started Guides](https://kb.ettus.com/X300/X310_Getting_Started_Guides)). The devices should be connected to single or different host computers based on the API operation mode and the investigated application. The following figure shows the setup of three Tx stations and one Rx Station.   
+To use the NI RF Data Recording API, you need at least one NI RF USRP device (for Tx or Rx only operation mode). The API has been tested both on [NI USRP X310](https://www.ettus.com/all-products/x310-kit/) ([X300/X310 Getting Started Guides](https://kb.ettus.com/X300/X310_Getting_Started_Guides)) and [NI USRP X410](https://www.ettus.com/all-products/usrp-x410/) ([X410 Getting Started Guides](https://kb.ettus.com/USRP_X410_Getting_Started_Guide)). The devices should be connected to single or different host computers based on the API operation mode and the investigated application. The following figure shows the setup of three Tx stations and one Rx Station.
```
Comment (Abdo-Gaber) [inline README.md:82]: has been tested on both

File: README.md [hunk 2]
```diff
@@ -79,18 +79,19 @@ Despite being in the early phase of research for 6G, AI & ML appears to be becom
...
 | Item    					| Number	| 
 | ------------------------- | ----------|
 | Host PC (Linux Server)   	| 1 		|
-| X310 USRP Device    		| 4    		| 
+| X310 or X410 USRP Device    		| 4    		| 
 | SMA Cable  				| 4 		| 
 | 30 dB Attenuator  		| 1    		| 
 | 10 Gig Eth Cable (SFP+) *	| 4     	|
-***Note**: The USRP X310 can be connect to the host machine using different options. Look to [Network Connectivity Guide](https://files.ettus.com/manual/page_usrp_x3x0.html).
+***Note1**: The USRP X310 can be connect to the host machine using different options. Look to [Network Connectivity Guide](https://files.ettus.com/manual/page_usrp_x3x0.html#x3x0_getting_started_connectivity).
```
Comment (Abdo-Gaber) [inline README.md:93]: Please update note for X310 and X410 "can be connected"

File: README.md [hunk 3]
```diff
@@ -79,18 +79,19 @@ Despite being in the early phase of research for 6G, AI & ML appears to be becom
...
+***Note2**: The USRP X410 can be connect to the host machine using different options. Look to [Network Connectivity Guide](https://files.ettus.com/manual/page_usrp_x4xx.html#x4xx_getting_started_network_connectivity).
```
Comment (Abdo-Gaber) [inline README.md:94]: can be connected

**Review by Abdo-Gaber — COMMENTED** [review]
Please take a look to the given comments, I need to do a test.

File: build/docker/Dockerfile [hunk 4]
```diff
@@ -30,12 +31,12 @@ RUN cd workarea
...
 
 # let's build UHD from source (note that with Dockerfile current directory path is not kept in the next commands)
-RUN cd uhd && cd host && mkdir build && cd build && cmake .. && make && make test && make install
+RUN cd uhd && git checkout v4.3.0.0 && cd host && mkdir build && cd build && cmake .. && make && make test && make install
 
 # setup the library references
 RUN ldconfig
-ENV LD_LIBRARY_PATH=/usr/local/lib
-ENV PYTHONPATH=/usr/local/lib/python3.8/site-packages
+RUN echo "LD_LIBRARY_PATH=/usr/local/lib">>~/.bashrc
```
[the remark this answers is not in the document]
Comment (LawsonGu) [reply build/docker/Dockerfile:38]: Yes, but only on Docker installed on Windows OS, without on Docker installed on Linux or Mac OS. I think most people would use Docker from Windows.

File: src/rf_replay_data_transmitter_usrp_uhd.py [hunk 5]
```diff
@@ -405,10 +408,14 @@ def main():
...
         print("Press Ctrl+C to stop RF streaming")
+        list = ["\\", "|", "/", "—"]
         while stop_tx_signal_called == False:
-            time.sleep(0.1)  # sleep for 100ms
-        # Remove SIGINT handler
-        # signal.signal(signal.SIGINT, signal_dfl)
+            for i in range(0, 4):
+                index = i % 4
+                print("\rRF streaming {}".format(list[index]), end="")
+                time.sleep(0.25) # sleep for 250ms
```
[the remark this answers is not in the document]
Comment (LawsonGu) [reply src/rf_replay_data_transmitter_usrp_uhd.py:416]: Perviously, when there istransmitting, everything is stable just like dead. So I add an animation just like a running sign.
The 0.25 is for running an animation, but 0.1 is ok, makes the animation run faster.
Don't you think animation is necessary? If the answer is no, we could recover to sleep(0.1)

File: src/lib/run_rf_data_recorder.py [hunk 6]
```diff
@@ -110,25 +113,25 @@ def rf_data_recorder(rx_args, txs_args, general_config, rx_data_nbytes_que):
         # To reduce latency, the number of records is executed per each configuration
         if i == 0:
             print(f"Requesting RX Freq: {(rx_args.freq / 1e6)} MHz...")
-            rx_args.coerced_rx_freq = usrp.get_rx_freq()
+            rx_args.coerced_rx_freq = usrp.get_rx_freq(rx_args.channels[0])
```
[the remark this answers is not in the document]
Comment (Abdo-Gaber) [reply src/lib/run_rf_data_recorder.py:116]: That is great catch. Ok, we can keep it. Please write a comment close to this line. In the future, if we are going to extend the code to capture from multiple channels, we should update the meta-data also. We can read those coerced values in a loop based on the channels order.

File: src/lib/rf_data_recording_api_def.py [hunk 7]
```diff
@@ -377,10 +379,45 @@ def round_up_to_even(f):
                 args_out = args_in + ",master_clock_rate=184.32e6"
             else:
                 args_out = args_in + ",master_clock_rate=200e6"
+        
+        # Derive master clock rate for X410 USRP
+        if "X4" in usrp_mboard_id:
```
Comment (Abdo-Gaber) [inline src/lib/rf_data_recording_api_def.py:384]: Please make it as if....elif ..else
In the current implementation: If X3 ... if X4 ...else .... If the USRP is X310, its arg will be overwritten by the arg under else of X410

**Review by Abdo-Gaber — APPROVED** [review]
It looks good. I was not able to test it using X410 on my side, but I did using X310. However, since you did the test already on your side, we can go ahead and merge it to main.

**Comment by LawsonGu** [comment]
Thanks

**Pull request merged**